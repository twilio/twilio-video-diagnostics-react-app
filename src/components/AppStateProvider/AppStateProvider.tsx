import React, { createContext, useContext, useReducer, useEffect } from 'react';
import produce from 'immer';
import { VideoInputTest } from '@twilio/rtc-diagnostics';

export enum ActivePane {
  GetStarted,
  DeviceCheck,
  DeviceError,
  CameraTest,
  Connectivity,
  Quality,
  Results,
}

interface stateType {
  activePane: ActivePane;
  videoGranted: boolean;
  audioGranted: boolean;
  deviceError: null | Error;
  videoInputTestReport: null | VideoInputTest.Report;
}

type ACTIONTYPE =
  | { type: 'set-active-pane'; newActivePane: ActivePane }
  | { type: 'next-pane' }
  | { type: 'previous-pane' }
  | { type: 'set-devices'; devices: MediaDeviceInfo[] }
  | { type: 'set-device-error'; error: Error }
  | { type: 'set-video-test-report'; report: VideoInputTest.Report };

type AppStateContextType = {
  state: stateType;
  dispatch: React.Dispatch<ACTIONTYPE>;
};

export const initialState = {
  activePane: ActivePane.GetStarted,
  videoGranted: false,
  audioGranted: false,
  deviceError: null,
  videoInputTestReport: null,
};

export const AppStateContext = createContext<AppStateContextType>(null!);

export function useAppStateContext() {
  const context = useContext(AppStateContext);
  if (!context) {
    throw new Error('useAppStateContext must be used within an AppStateProvider');
  }
  return context;
}

/**
 * This reducer function helps with the logic used to decide the activePane,
 * as well as which panes are to be shown to the user, and when.
 */

export const appStateReducer = produce((draft: stateType, action: ACTIONTYPE) => {
  switch (action.type) {
    case 'set-active-pane':
      draft.activePane = action.newActivePane;
      break;

    case 'next-pane':
      switch (draft.activePane) {
        case ActivePane.GetStarted:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.CameraTest;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        case ActivePane.DeviceCheck:
          if (draft.deviceError) {
            draft.activePane = ActivePane.DeviceError;
          } else {
            draft.activePane = ActivePane.CameraTest;
          }
          break;
        default:
          draft.activePane++;
          break;
      }
      break;

    case 'previous-pane':
      switch (draft.activePane) {
        case ActivePane.CameraTest:
          if (draft.audioGranted && draft.videoGranted) {
            draft.activePane = ActivePane.GetStarted;
          } else {
            draft.activePane = ActivePane.DeviceCheck;
          }
          break;
        default:
          draft.activePane--;
          break;
      }
      break;

    case 'set-devices':
      draft.audioGranted = action.devices.filter((d) => d.kind === 'audioinput').every((d) => d.label);
      draft.videoGranted = action.devices.filter((d) => d.kind === 'videoinput').every((d) => d.label);
      break;

    case 'set-device-error':
      draft.deviceError = action.error;
      draft.activePane = ActivePane.DeviceError;
      break;

    case 'set-video-test-report':
      draft.videoInputTestReport = action.report;
  }
});

export const AppStateProvider: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(appStateReducer, initialState);

  useEffect(() => {
    navigator.mediaDevices.enumerateDevices().then((devices) => dispatch({ type: 'set-devices', devices }));
  }, []);

  return <AppStateContext.Provider value={{ state, dispatch }}>{children}</AppStateContext.Provider>;
};
