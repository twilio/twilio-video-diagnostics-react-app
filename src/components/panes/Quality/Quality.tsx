import { ExcellentQuality } from './ExcellentQuality/ExcellentQuality';
import { PoorQuality } from './PoorQuality/PoorQuality';
import { getQualityScore } from './getQualityScore/getQualityScore';
import { useAppStateContext } from '../../AppStateProvider/AppStateProvider';
import { QualityModal } from './QualityModal/QualityModal';
import { useState } from 'react';

export enum QualityScore {
  Poor,
  Suboptimal,
  Good,
  Excellent,
}

export function Quality() {
  const { state } = useAppStateContext();
  const [isModalOpen, setIsModalOpen] = useState(false);
  const { totalQualityScore, latency, jitter, packetLoss, bitrate } = getQualityScore(
    state.preflightTest.report,
    state.bitrateTest.report
  );

  return (
    <>
      <QualityModal
        isModalOpen={isModalOpen}
        setIsModalOpen={setIsModalOpen}
        jitter={jitter}
        latency={latency}
        packetLoss={packetLoss}
        bitrate={bitrate}
      />
      {totalQualityScore === QualityScore.Good || totalQualityScore === QualityScore.Excellent ? (
        <ExcellentQuality quality={totalQualityScore} openModal={() => setIsModalOpen(true)} />
      ) : (
        <PoorQuality quality={totalQualityScore} openModal={() => setIsModalOpen(true)} />
      )}
    </>
  );
}
