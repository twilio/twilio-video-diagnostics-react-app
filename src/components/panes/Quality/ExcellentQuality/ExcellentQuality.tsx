import { Container, Typography, Grid, Button, makeStyles } from '@material-ui/core';
import { useAppStateContext } from '../../../AppStateProvider/AppStateProvider';
import { QualityScore } from '../Quality';
import Excellent from './Excellent.png';
import { ViewIcon } from '../../../../icons/ViewIcon';

const useStyles = makeStyles({
  modal: {
    width: '6000px',
  },
  illustrationContainer: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  viewButton: {
    marginTop: '2em',
    '& svg': {
      position: 'relative',
      left: '-5px',
    },
  },
});

interface ExcellentQualityProps {
  quality: QualityScore;
  openModal: () => void;
}

export function ExcellentQuality({ quality, openModal }: ExcellentQualityProps) {
  const { nextPane } = useAppStateContext();
  const classes = useStyles();

  const qualityScore = quality === QualityScore.Excellent ? 'excellent' : 'good';

  return (
    <Container>
      <Grid container alignItems="center" justifyContent="space-between">
        <Grid item md={5}>
          <Typography variant="h1" gutterBottom>
            Expected quality is {qualityScore}
          </Typography>

          <Typography variant="body1" gutterBottom>
            This is the last step! Your expected audio and video quality is <strong>{qualityScore}</strong> and overall
            performance looks {qualityScore === 'excellent' ? 'good' : 'ok'}.
          </Typography>

          <Button variant="contained" color="primary" onClick={nextPane}>
            Finish up!
          </Button>
        </Grid>

        <Grid item md={5} className={classes.illustrationContainer}>
          {/* 
      The size of the image is explicitly stated here so that this content can properly be centered vertically
      before the image is loaded.
      */}
          <img src={Excellent} alt="Success" style={{ width: '245px', height: '200px' }} />
          <Button variant="outlined" onClick={openModal} className={classes.viewButton}>
            <ViewIcon />
            View detailed quality information
          </Button>
        </Grid>
      </Grid>
    </Container>
  );
}
