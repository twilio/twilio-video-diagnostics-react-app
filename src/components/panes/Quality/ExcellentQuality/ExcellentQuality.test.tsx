import { shallow } from 'enzyme';
import { useAppStateContext } from '../../../AppStateProvider/AppStateProvider';
import { QualityScore } from '../Quality';
import { ExcellentQuality } from './ExcellentQuality';

jest.mock('../../../AppStateProvider/AppStateProvider');

const mockUseAppStateContext = useAppStateContext as jest.Mock<any>;

mockUseAppStateContext.mockImplementation(() => ({ nextPane: jest.fn() }));

const mockSetIsModalOpen = jest.fn();

describe('the Excellent Quality component', () => {
  it('should render correctly when quality is "good"', () => {
    const wrapper = shallow(<ExcellentQuality quality={QualityScore.Good} openModal={mockSetIsModalOpen} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when quality is "excellent"', () => {
    const wrapper = shallow(<ExcellentQuality quality={QualityScore.Excellent} openModal={mockSetIsModalOpen} />);
    expect(wrapper).toMatchSnapshot();
  });
});
