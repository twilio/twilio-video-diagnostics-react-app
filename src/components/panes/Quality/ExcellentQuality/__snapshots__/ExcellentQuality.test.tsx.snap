// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Excellent Quality component should render correctly when quality is "excellent" 1`] = `
<WithStyles(ForwardRef(Container))>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    justifyContent="space-between"
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      md={5}
    >
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="h1"
      >
        Quality is 
        excellent
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="body1"
      >
        This is the last step! Your expected audio and video quality is 
        <strong>
          excellent
        </strong>
         and overall performance looks 
        good
        .
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[MockFunction]}
        variant="contained"
      >
        Finish up!
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-illustrationContainer-2"
      item={true}
      md={5}
    >
      <img
        alt="Success"
        src="test-file-stub"
        style={
          Object {
            "height": "200px",
            "width": "245px",
          }
        }
      />
      <WithStyles(ForwardRef(Button))
        className="makeStyles-viewButton-3"
        onClick={[MockFunction]}
        variant="outlined"
      >
        <ViewIcon />
        View detailed quality information
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Container))>
`;

exports[`the Excellent Quality component should render correctly when quality is "good" 1`] = `
<WithStyles(ForwardRef(Container))>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    container={true}
    justifyContent="space-between"
  >
    <WithStyles(ForwardRef(Grid))
      item={true}
      md={5}
    >
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="h1"
      >
        Quality is 
        good
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        gutterBottom={true}
        variant="body1"
      >
        This is the last step! Your expected audio and video quality is 
        <strong>
          good
        </strong>
         and overall performance looks 
        ok
        .
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[MockFunction]}
        variant="contained"
      >
        Finish up!
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      className="makeStyles-illustrationContainer-2"
      item={true}
      md={5}
    >
      <img
        alt="Success"
        src="test-file-stub"
        style={
          Object {
            "height": "200px",
            "width": "245px",
          }
        }
      />
      <WithStyles(ForwardRef(Button))
        className="makeStyles-viewButton-3"
        onClick={[MockFunction]}
        variant="outlined"
      >
        <ViewIcon />
        View detailed quality information
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(Grid))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Container))>
`;
